yt/yt/server/master/object_server/object_service.cpp

```diff
@@ -1600,9 +1603,20 @@ class TObjectService::TExecuteSession
                 }
 
                 const auto& securityManager = Bootstrap_->GetSecurityManager();
-                auto result = securityManager->ThrottleUser(User_.Get(), 1, workloadType);
 
-                if (!WaitForAndContinue(result)) {
+                auto automatonThrottlerFuture = VoidFuture;
+                if (subrequestType == EExecutionSessionSubrequestType::LocalWrite && User_.Get() != securityManager->GetRootUser()) {
+                    automatonThrottlerFuture = Owner_->AutomatonRequestThrottler_->Throttle(1);
+                }
+
+                // NB: This callback may be executed in an arbitrary thread.
```

Comment:
It is definitely fragile, but not sure how avoidable that is.

Re arbitrary thread: we either rely on `securityManager->ThrottleUser` being thread-safe (which, in the end, it [is](https://github.com/ytsaurus/ytsaurus/blob/main/yt/yt/server/master/security_server/user.cpp#L462)), or we drag this callback through the current invoker (besides `Automaton` this can also be a `LocalRead` thread), which isn't that efficient. 

User pointer should not be dead because we capture a strong ref to `this`, which contains `TEphemeralObjectPtr<TUser> User_`. I asked Grisha about it, and he told me that it should guarantee that the user is alive. Maybe I misinterpreted something, is this wrong? 
The problem with capturing `User_` as a whole is that `Get()` contains `VerifyPersistentStateRead`, which would (once again) require us to run explicitly in Automaton/LocalRead threads.

Since it is obviously a concern, I want to hear your thoughts on how to do all of this best.

Reply:
Let's break it down.

1. `securityManager->ThrottleUser` being thread-safe is fine and I don't see a better way to express in code what we're doing here.


2. The point you're making about strong-capturing `this` and, indirectly, preventing destruction of the user object if flawed, sadly. What we're asctually doing is prolonging the lifetime of the *ephemeral pointer.* That pointer, however, does not prevent the user object from being destroyed. An ephemeral pointer is, semantically, a weak one: it prolongs the lifetime of the control block but not the whole object. Thus, right before dereferencing the pointer, one has to check the refcounters to determine whether the object is still alive. Here's the sad part: those refcounters may only be read from persistent-state-read threads (i.e. either the Automaton or LocalRead).

    Basically, this is the long-winded way of saying "unless you strong-ref the user, checking its liveness must be done from persistent-state-read threads only". (And strong-reffing the user is a no-go since that would delay the destruction of a removed used indefinitely - for as long as it continues issuing requests.)

3. There's a discussion to be had about what could be expressed by the question "shouldn't we call `::Throttle` on both the global and per-user throttlers simultaneously instead of sequentially"? It appears you've had this discussion with Grigory but I just wanted to note that changing you mind on this *could* solve the issue.

Reply:
Coming back to this, I think the simplest way is to reconsider calling the throttlers sequentially vs simultaneously. There is the annoying issue of keeping the underlying promise within AllSet alive, but I guess there is no smart way around it and I'll just store the promise in the session object. I will make a fast path when both futures are already set though.

Reply:
Ah, I remember now, the callback we subscribe with has a strong ref to TExecuteSession, so we cannot store the promise inside the session. 😞

Reply:
Not sure why I didn't think of this sooner, but we can get both futures and chain them with an Apply as a poor-man's replacement for AllSet. This way nobody has to own a new promise. The underlying promises of both throttler futures are referenced by the corresponding throttlers and everyone is happy this way.